File: integrations/PaystackOSS/paystack_flutter/android/src/main/kotlin/com/paystack/paystack_sdk/PaystackSDKPlugin.kt
package com.paystack.paystack_sdk

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import com.paystack.android.core.Paystack
import com.paystack.android.ui.paymentsheet.PaymentSheet
import com.paystack.android.ui.paymentsheet.PaymentSheetResult
import io.flutter.embedding.android.FlutterFragmentActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.embedding.engine.plugins.lifecycle.HiddenLifecycleReference
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class PaystackSDKPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  private lateinit var channel : MethodChannel
  private lateinit var paymentSheet: PaymentSheet
  private var activity: FlutterFragmentActivity? = null
  private lateinit var pendingResult: Result

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.paystack.flutter")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    pendingResult = result
    when (call.method) {
      "initialize" -> {
        val publicKey = call.argument<String>("publicKey")
        val enableLogging = call.argument<Boolean>("enableLogging") ?: false
        if (publicKey.isNullOrEmpty()) {
          result.error(
            "INVALID_ARGUMENT",
            "Missing public key",
            null
          )
          return
        }
        initialize(publicKey, enableLogging)
      }
      "launch" -> {
        val accessCode = call.argument<String>("accessCode")
        if(accessCode.isNullOrEmpty()) {
          result.error(
            "INVALID_ARGUMENT",
            "Missing access code",
            null
          )
          return
        }
        launch(accessCode)
      }
      else -> result.notImplemented()
    }
  }

  private fun initialize(publicKey: String, enableLogging: Boolean) {
    try {
      Paystack
        .builder()
        .setPublicKey(publicKey)
        .setLoggingEnabled(enableLogging)
        .build()
      pendingResult.success(true)
    } catch (e: Exception) {
      pendingResult.error(
        "INITIALIZATION_ERROR",
        e.message,
        null
      )
    }
  }

  private fun launch(accessCode: String) {
    if (activity == null) {
      pendingResult.error(
        "MISSING_VIEW",
        "Activity is not found to present payment UI",
        null
      )
      return
    }

    try {
      paymentSheet.launch(accessCode)
    } catch (e: Exception) {
      pendingResult.error("LAUNCH_ERROR", e.message, null)
      return
    }
  }

  private fun paymentComplete(paymentSheetResult: PaymentSheetResult) {
    when (paymentSheetResult) {
      is PaymentSheetResult.Completed -> {
        pendingResult.success(mapOf(
          "status" to "success",
          "message" to "Transaction successful",
          "reference" to paymentSheetResult.paymentCompletionDetails.reference
        ))
      }
      is PaymentSheetResult.Cancelled -> {
        pendingResult.success(mapOf(
          "status" to "cancelled",
          "message" to "Transaction cancelled",
          "reference" to ""
        ))
      }
      is PaymentSheetResult.Failed -> {
        pendingResult.success(mapOf(
          "status" to "failed",
          "message" to paymentSheetResult.error.message,
          "reference" to ""
        ))
      }
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity as FlutterFragmentActivity
    (binding.lifecycle as HiddenLifecycleReference)
      .lifecycle
      .addObserver(LifecycleEventObserver { _, event ->
        if (event == Lifecycle.Event.ON_CREATE) {
          val activity = requireNotNull(activity)
          paymentSheet = PaymentSheet(activity, ::paymentComplete)
        }
      })
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity as FlutterFragmentActivity
  }

  override fun onDetachedFromActivity() {
    activity = null
  }
}

File: integrations/PaystackOSS/paystack_flutter/ios/paystack_flutter_sdk/Sources/paystack_sdk/PaystackSDKPlugin.swift
import Flutter
import UIKit
import PaystackCore
import PaystackUI

public class PaystackSDKPlugin: NSObject, FlutterPlugin {
    private var paystack: Paystack?
    private var result: FlutterResult?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.paystack.flutter", binaryMessenger: registrar.messenger())
        let instance = PaystackSDKPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        self.result = result
        switch call.method {
        case "initialize":
            guard let args = call.arguments as? [String: Any],
                let publicKey = args["publicKey"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT",
                                    message: "Missing public key",
                                    details: nil))
                return
            }
            let logging = args["enableLogging"] as? Bool ?? false
            initialize(publicKey: publicKey, logging: logging)
        case "launch":
            guard let args = call.arguments as? [String: Any],
                let accessCode = args["accessCode"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT",
                                    message: "Missing access code",
                                    details: nil))
                return
            }
            launch(accessCode: accessCode)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initialize(publicKey: String, logging: Bool) {
        do {
            let instance = PaystackBuilder
                .newInstance
                .setKey(publicKey)

            if logging {
                instance.enableLogging()
            }
            
            paystack = try instance.build()
            self.result?(true)
        } catch {
            self.result?(FlutterError(code: "INITIALIZATION_ERROR",
                                 message: error.localizedDescription,
                                 details: nil))
        }
    }
    
    private func launch(accessCode: String) {
        guard let paystack = paystack else {
            self.result?(FlutterError(code: "INITIALIZATION_ERROR",
                                 message: "Paystack not initialized",
                                details: nil))
            return
        }
        
        guard #available(iOS 14.0, *) else {
            self.result?(FlutterError(code: "UNSUPPORTED_VERSION",
                                 message: "Paystack UI requires iOS 14.0 or later",
                                details: nil))
            return
        }
        
        guard let viewController = UIApplication.shared.keyWindow?.rootViewController else {
            self.result?(FlutterError(code: "MISSING_VIEW",
                                 message: "View controller not found to present payment UI",
                                details: nil))
            return
        }
        
        paystack.presentChargeUI(
            on: viewController,
            accessCode: accessCode,
            onComplete: { [weak self] transactionResult in
                self?.paymentDone(transactionResult)
        })
    }
    
    private func paymentDone(_ transactionResult: TransactionResult) {
       switch (transactionResult) {
         case .completed(let details):
           self.result?([
                "status": "success",
                "message": "Transaction successful",
                "reference": details.reference
           ])
         case .cancelled:
           self.result?([
                "status": "cancelled",
                "message": "Transaction cancelled",
                "reference": ""
           ])
         case .error(error: let error, reference: let reference):
           self.result?([
                "status": "failed",
                "message": error.message,
                "reference": ""
           ])
       }
     }
}
